README
NAME
    Thread::Cleanup - Hook thread destruction.

VERSION
    Version 0.06

SYNOPSIS
        use Thread::Cleanup;

        use threads;

        Thread::Cleanup::register {
         my $tid = threads->tid();
         warn "Thread $tid finished\n";
        };

DESCRIPTION
    This module allows you to hook thread destruction without fiddling with
    the internals of threads.

    It acts globally on all the threads that may spawn anywhere in your
    program, with the exception of the main thread.

    The hook will also be called when pseudo-forks (i.e. processes spawn on
    Windows for the "fork" emulation) terminate.

FUNCTIONS
  "register"
        register { ... };
        &register($coderef);

    Specify that the given block or code reference $coderef will have to be
    called (in void context, without arguments) every time a thread or a
    pseudo-fork terminates. More precisely :

    *   For joined threads, it will be called when "join" succeeds, after
        any "END" block local to the spawn thread ;

    *   For detached threads, it will be called if and only if the thread
        terminates before the main thread, and the hook will then fire at
        global "END" time ;

    *   For pseudo-forks, it will be called when "waitpid" succeeds, after
        any "END" block local to the spawn process but before any global
        "END" block ;

    *   It will never trigger for the destruction of the main thread.

EXPORT
    None.

DEPENDENCIES
    perl 5.8.

    A C compiler. This module may happen to build with a C++ compiler as
    well, but don't rely on it, as no guarantee is made in this regard.

    threads 1.07.

    XSLoader.

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

BUGS
    Please report any bugs or feature requests to "bug-thread-cleanup at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Thread-Cleanup>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Thread::Cleanup

ACKNOWLEDGEMENTS
    Inspired by a question from TonyC on #p5p.

COPYRIGHT & LICENSE
    Copyright 2009,2010,2013,2014 Vincent Pit, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Thread-Cleanup

0.06    2014-09-28 15:40 UTC
        + Doc : Updates and clarifications.
        + Fix : The hooks registered by this module will now be reliably
                triggered when pseudo-forks (e.g. processes spawn on Windows
                for the fork() emulation) terminate.

0.05    2013-09-01 16:50 UTC
        This is a maintenance release. The code contains no functional change.
        Satisfied users of version 0.04 can skip this update.
        + Doc : C++ compilers are officially NOT supported.
        + Doc : POD headings are now properly linkable.
        + Tst : Author tests are no longer bundled with this distribution.
                They are only made available to authors in the git repository.
        + Tst : Threads tests will not fail anymore if resources constraints
                prevent the system from creating all the required threads.

0.04    2010-05-31 22:55 UTC
        + Fix : Compatibility with perl 5.13.1.

0.03    2010-01-06 23:30 UTC
        + Fix : Named scopes with debugging blead.
        + Fix : Work around Kwalitee test misfailures.

0.02    2009-06-04 21:35 UTC
        + Doc : Typos and nits.
        + Fix : Dieing inside a Thread::Cleanup callback no longer causes the
                whole process to exit.
        + Tst : Test that detached threads destructors don't fire before END.

0.01    2009-03-16 00:30 UTC
        First version, released on an unsuspecting world.



reap.h
/* This file is part of the Thread::Cleanup Perl module.
 * See http://search.cpan.org/dist/Thread-Cleanup/ */

/* This header provides a specialized version of Scope::Upper::reap that can be
 * called directly from XS.
 * See http://search.cpan.org/dist/Scope-Upper/ for details. */

#ifndef REAP_H
#define REAP_H 1

#define REAP_DESTRUCTOR_SIZE 3

typedef struct {
 I32    depth;
 I32   *origin;
 void (*cb)(pTHX_ void *);
 void  *ud;
 char  *dummy;
} reap_ud;

STATIC void reap_pop(pTHX_ void *);

STATIC void reap_pop(pTHX_ void *ud_) {
 reap_ud *ud = ud_;
 I32 depth, *origin, mark, base;

 depth  = ud->depth;
 origin = ud->origin;
 mark   = origin[depth];
 base   = origin[depth - 1];

 if (base < mark) {
  PL_savestack_ix = mark;
  leave_scope(base);
 }
 PL_savestack_ix = base;

 if ((ud->depth = --depth) > 0) {
  SAVEDESTRUCTOR_X(reap_pop, ud);
 } else {
  void (*cb)(pTHX_ void *) = ud->cb;
  void  *cb_ud             = ud->ud;

  PerlMemShared_free(ud->origin);
  PerlMemShared_free(ud);

  SAVEDESTRUCTOR_X(cb, cb_ud);
 }
}

STATIC void reap(pTHX_ I32 depth, void (*cb)(pTHX_ void *), void *cb_ud) {
#define reap(D, CB, UD) reap(aTHX_ (D), (CB), (UD))
 reap_ud *ud;
 I32 i;

 if (depth > PL_scopestack_ix)
  depth = PL_scopestack_ix;

 ud         = PerlMemShared_malloc(sizeof *ud);
 ud->depth  = depth;
 ud->origin = PerlMemShared_malloc((depth + 1) * sizeof *ud->origin);
 ud->cb     = cb;
 ud->ud     = cb_ud;
 ud->dummy  = NULL;

 for (i = depth; i >= 1; --i) {
  I32 j = PL_scopestack_ix - i;
  ud->origin[depth - i] = PL_scopestack[j];
  PL_scopestack[j] += REAP_DESTRUCTOR_SIZE;
 }
 ud->origin[depth] = PL_savestack_ix;

 while (PL_savestack_ix + REAP_DESTRUCTOR_SIZE
                                       <= PL_scopestack[PL_scopestack_ix - 1]) {
  save_pptr(&ud->dummy);
 }

 SAVEDESTRUCTOR_X(reap_pop, ud);
}

#endif /* REAP_H */


lib/Thread/Cleanup.pm
package Thread::Cleanup;

use 5.008;

use strict;
use warnings;

=head1 NAME

Thread::Cleanup - Hook thread destruction.

=head1 VERSION

Version 0.06

=cut

our $VERSION;

BEGIN {
 $VERSION = '0.06';
 require XSLoader;
 XSLoader::load(__PACKAGE__, $VERSION);
}

=head1 SYNOPSIS

    use Thread::Cleanup;

    use threads;

    Thread::Cleanup::register {
     my $tid = threads->tid();
     warn "Thread $tid finished\n";
    };

=head1 DESCRIPTION

This module allows you to hook thread destruction without fiddling with the internals of L<threads>.

It acts globally on all the threads that may spawn anywhere in your program, with the exception of the main thread.

The hook will also be called when pseudo-forks (i.e. processes spawn on Windows for the C<fork> emulation) terminate.

=head1 FUNCTIONS

=head2 C<register>

    register { ... };
    &register($coderef);

Specify that the given block or code reference C<$coderef> will have to be called (in void context, without arguments) every time a thread or a pseudo-fork terminates.
More precisely :

=over 4

=item *

For joined threads, it will be called when C<join> succeeds, after any C<END> block local to the spawn thread ;

=item *

For detached threads, it will be called if and only if the thread terminates before the main thread, and the hook will then fire at global C<END> time ;

=item *

For pseudo-forks, it will be called when C<waitpid> succeeds, after any C<END> block local to the spawn process but before any global C<END> block ;

=item *

It will never trigger for the destruction of the main thread.

=back

=cut

my @callbacks;

sub register (&) { push @callbacks, shift }

sub _CLEANUP { $_->() for @callbacks }

=head1 EXPORT

None.

=head1 DEPENDENCIES

L<perl> 5.8.

A C compiler.
This module may happen to build with a C++ compiler as well, but don't rely on it, as no guarantee is made in this regard.

L<threads> 1.07.

L<XSLoader>.

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

=head1 BUGS

Please report any bugs or feature requests to C<bug-thread-cleanup at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Thread-Cleanup>.  I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Thread::Cleanup

=head1 ACKNOWLEDGEMENTS

Inspired by a question from TonyC on #p5p.

=head1 COPYRIGHT & LICENSE

Copyright 2009,2010,2013,2014 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1; # End of Thread::Cleanup


Makefile.PL
use 5.008;

use strict;
use warnings;
use ExtUtils::MakeMaker;

my $dist = 'Thread-Cleanup';

(my $name = $dist) =~ s{-}{::}g;

(my $file = $dist) =~ s{-}{/}g;
$file = "lib/$file.pm";

my %PREREQ_PM = (
 'XSLoader' => 0,
 'threads'  => '1.07',
);

my %META = (
 configure_requires => {
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  'ExtUtils::MakeMaker' => 0,
  'Test::More'          => 0,
  'threads::shared'     => '0.91',
  %PREREQ_PM,
 },
 dynamic_config => 0,
 resources => {
  bugtracker => "http://rt.cpan.org/NoAuth/ReportBug.html?Queue=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
 NAME             => $name,
 AUTHOR           => 'Vincent Pit <perl@profvince.com>',
 LICENSE          => 'perl',
 VERSION_FROM     => $file,
 ABSTRACT_FROM    => $file,
 PL_FILES         => {},
 PREREQ_PM        => \%PREREQ_PM,
 MIN_PERL_VERSION => 5.008,
 META_MERGE       => \%META,
 dist             => {
  PREOP    => "pod2text $file > \$(DISTVNAME)/README",
  COMPRESS => 'gzip -9f', SUFFIX => 'gz'
 },
 clean            => {
  FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt"
 }
);
